### PR TITLE
TYP: Add DatetimeIndex.intersection() method and test 

### LINKED ISSUE #744: Index.intersection does not return correct sub type
**Describe the bug**
When I form the `intersection` of two `pd.DatetimeIndex`, the return type is `Index`. This leads to problems with index entries: mypy complains about `intersection[0].date()`.

**To Reproduce**
1. Minimal example
```py
import pandas as pd
intersection = pd.DatetimeIndex().intersection(pd.DatetimeIndex)

reveal_type(intersection)  # Index
reveal_type(intersection[0])  # "str | bytes | date | datetime | timedelta | datetime64 | timedelta64 | bool | int | float | Timestamp | Timedelta | complex"
```

2. mypy

3. error message

```py
agri_data/utils.py:96: error: Item "str" of "str | bytes | date | datetime | timedelta | datetime64 | timedelta64 | bool | int | float | Timestamp | Timedelta | complex" has no attribute "date"  [union-attr]
agri_data/utils.py:96: error: Item "bytes" of "str | bytes | date | datetime | timedelta | datetime64 | timedelta64 | bool | int | float | Timestamp | Timedelta | complex" has no attribute "date"  [union-attr]
agri_data/utils.py:96: error: Item "date" of "str | bytes | date | datetime | timedelta | datetime64 | timedelta64 | bool | int | float | Timestamp | Timedelta | complex" has no attribute "date"  [union-attr]
agri_data/utils.py:96: error: Item "timedelta" of "str | bytes | date | datetime | timedelta | datetime64 | timedelta64 | bool | int | float | Timestamp | Timedelta | complex" has no attribute "date"  [union-attr]
agri_data/utils.py:96: error: Item "datetime64" of "str | bytes | date | datetime | timedelta | datetime64 | timedelta64 | bool | int | float | Timestamp | Timedelta | complex" has no attribute "date"  [union-attr]
agri_data/utils.py:96: error: Item "timedelta64" of "str | bytes | date | datetime | timedelta | datetime64 | timedelta64 | bool | int | float | Timestamp | Timedelta | complex" has no attribute "date"  [union-attr]
agri_data/utils.py:96: error: Item "bool" of "str | bytes | date | datetime | timedelta | datetime64 | timedelta64 | bool | int | float | Timestamp | Timedelta | complex" has no attribute "date"  [union-attr]
agri_data/utils.py:96: error: Item "int" of "str | bytes | date | datetime | timedelta | datetime64 | timedelta64 | bool | int | float | Timestamp | Timedelta | complex" has no attribute "date"  [union-attr]
agri_data/utils.py:96: error: Item "float" of "str | bytes | date | datetime | timedelta | datetime64 | timedelta64 | bool | int | float | Timestamp | Timedelta | complex" has no attribute "date"  [union-attr]
agri_data/utils.py:96: error: Item "Timedelta" of "str | bytes | date | datetime | timedelta | datetime64 | timedelta64 | bool | int | float | Timestamp | Timedelta | complex" has no attribute "date"  [union-attr]
agri_data/utils.py:96: error: Item "complex" of "str | bytes | date | datetime | timedelta | datetime64 | timedelta64 | bool | int | float | Timestamp | Timedelta | complex" has no attribute "date"  [union-attr]
```

4. Expected type: `reveal_type(intersection[0])  # Timestamp`

**Please complete the following information:**
 - OS: Ubuntu
 - OS Version: bullseye
 - python version: 3.11.1
 - version of type checker: 1.2.0
 - version of installed `pandas-stubs`: 2.0.2.230605

**Additional context**
Possibly related: #723

### PR DESCRIPTION
**Pull request by teacherc**
Fix DatetimeIndex.intersection() return type

The return type of DatetimeIndex.intersection() was incorrectly 
specified.This updates the method signature to indicate the correct return 
type of DatetimeIndex.

Test added as well.

Fixes #744 

- [X] Closes #xxxx (Replace xxxx with the Github issue number)
- [X] Tests added: Please use `assert_type()` to assert the type of any return value

### CODE REVIEW

File: pandas-stubs/core/indexes/datetimes.pyi [hunk 1]
```diff
@@ -97,6 +98,7 @@ class DatetimeIndex(  # type: ignore[misc]
     def tzinfo(self) -> tzinfo | None: ...
     @property
     def dtype(self) -> np.dtype | DatetimeTZDtype: ...
+    def intersection(self, other: Index | list) -> DatetimeIndex: ...
```
Comment (Dr-Irv) [inline pandas-stubs/core/indexes/datetimes.pyi:101]: The type of `other` should be `DatetimeIndex` .

File: tests/test_indexes.py [hunk 2]
```diff
@@ -978,3 +978,12 @@ def test_index_constructors():
...
         pd.Index(flist, dtype=np.float16)
+
+def test_datetimeindex_intersection():
+
+    idx1 = pd.DatetimeIndex(['2022-01-01', '2022-01-02']) 
+    idx2 = pd.DatetimeIndex(['2022-01-02', '2022-01-03'])
+
+    result = idx1.intersection(idx2)
+
+    assert_type(result, pd.DatetimeIndex)
```
Comment (Dr-Irv) [inline tests/test_indexes.py:989]: can you write this as `check(assert_type(result, pd.DatetimeIndex), pd.DatetimeIndex)`

**Review by Dr-Irv — CHANGES_REQUESTED** [review]

File: tests/test_indexes.py (hunk 2 above)
Comment (teacherc) [reply tests/test_indexes.py]: Done. Thanks!

File: pandas-stubs/core/indexes/datetimes.pyi (hunk 1 above)
Comment (teacherc) [reply pandas-stubs/core/indexes/datetimes.pyi]: Got it.

File: pandas-stubs/core/indexes/datetimes.pyi [hunk 3]
```diff
@@ -97,6 +97,7 @@ class DatetimeIndex(  # type: ignore[misc]
     def tzinfo(self) -> tzinfo | None: ...
     @property
     def dtype(self) -> np.dtype | DatetimeTZDtype: ...
+    def intersection(self, other: DatetimeIndex | list) -> DatetimeIndex: ...
```
Comment (Dr-Irv) [inline pandas-stubs/core/indexes/datetimes.pyi:100]: `list` is too broad.  I think `Sequence[DatetimeLike]` will work where you import `DatetimeLike` from `pandas._typing`

File: tests/test_indexes.py [hunk 4]
```diff
@@ -978,3 +978,12 @@ def test_index_constructors():
...
         pd.Index(flist, dtype=np.float16)
+
+
+def test_datetimeindex_intersection():
+    idx1 = pd.DatetimeIndex(["2022-01-01", "2022-01-02"])
+    idx2 = pd.DatetimeIndex(["2022-01-02", "2022-01-03"])
+
+    result = idx1.intersection(idx2)
+
+    check(assert_type(result, pd.DatetimeIndex), pd.DatetimeIndex)
```
Comment (Dr-Irv) [inline tests/test_indexes.py:989]: If you include the `list[DatetimeLike]` as I suggested, then you should include tests that do `intersection()` with lists of those 3 different types.  E.g., `idx1.intersection([pd.Timestamp(x) for x in idx2])` and something similar for `dt.datetime` and `np.datetime64`

**Review by Dr-Irv — CHANGES_REQUESTED** [review]

**Comment by twoertwein** [comment]
I believe this has been addressed as part of #760 . Feel free to re-open this PR if #760 didn't cover all aspects.

**Pull request closed without merging**